### PR TITLE
multiprocessing support for hicBuildMatrix

### PR DESCRIPTION
**Pull request by joachimwolff**
This PR adds a multiprocessing support to build the hic matrix. It uses pythons multiprocessing library (https://docs.python.org/2/library/multiprocessing.html), shared memory resources are implemented with shared ctypes (https://docs.python.org/2/library/multiprocessing.html#module-multiprocessing.sharedctypes). 
The parallelization is splittet in two parts: One main process reads the two input bam files and filters for duplicates. After a predefined buffer size is reached, the instances are passed to one process where the rest of the computation to create the matrix is done. The main process reads now the next 'bufferSize'-elements and passes these two a second process etc. This model limits the number of useful processes. If the I/O main process is not fast enough to read the data in the buffer for the n-th process by the time the first process is finished with the computation, then adding more processes will not speed up the computation anymore.
Each process writes its own output.bam file to a predefined buffer dir (parameter: --outputFileBufferDir). Default is '/dev/shm/' which is in the most Linux systems an easy to use ram disk. The subfiles of each process are merged in the background with an additional process. After the computation is done, the output.bam file is copied to to defined location 'outFileName' which is usually the local folder. 
Overall the number of processes with the parameter '--threads' plus two processes (reading input bam + merging output bam) are running in the computation. 
Pythons multiprocessing library copies all data needed for a process which causes a memory overhead. Possible solutions with a multiprocessing.Manager to share the needed resources were tested but caused a slowdown of the computation. To overcome this the data structure for  'bin_intval_tree' and 'coverage' are changed to a 1D multiprocessing.sharedctypes Array with a specific 'struct' for each of them. This caused some changes in the computation e.g. the search for the right position in the hic matrix with the bin_intval_tree is now implemented with a binary search. 

The dataset used for testing is: ```ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR195/007/SRR1956527/SRR1956527_2.fastq.gz```
```ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR195/007/SRR1956527/SRR1956527_1.fastq.gz```

The speedup with the SRR1956527 data set is around 2 - 3; 20,000 - 25,000 lines a second vs. 55 - 71,000 lines a second; ~54 min runtime vs. ~30 min; ~10 GB vs. ~15 GB RAM usage. This was computed with 4 + 2 processes, a bin size of 10,000 and a buffer size of 200,000. 
The runtime can be decreased with a higher number of processes and a bigger buffer size which results in a higher memory usage. E.g. with 5 + 2 processes and a buffer size of 400,000 around 70,000 - 95,000 lines a second are possible; ~23 min runtime; ~22 GB RAM was used. This was computed with NFS storage.
The limitation is the I/O. With the usage of a SSD 80,000 - 100,000 lines a second ( 4 + 2 processes, 200,000 buffer size) are possible and I don't think that this is the upper limit but I don't have more cores to test it on my notebook.

### CODE REVIEW

**Comment by bgruening** [comment]
Nice!!!

File: hicexplorer/hicBuildMatrix.py [hunk 1]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
-            vec_end = min(len(coverage[mate_bin_id]), vec_start +
-                          len(mate.seq) / binsize)
-            coverage[mate_bin_id][vec_start:vec_end] += 1
+            vec_start = max(0, mate.pos - mate_bin.begin) / pBinsize
+            length_coverage = pCoverageIndex[mate_bin_id].end - pCoverageIndex[mate_bin_id].begin
+            vec_end = min(length_coverage, vec_start +
+                          len(mate.seq) / pBinsize)
+            coverage_index = pCoverageIndex[mate_bin_id].begin + vec_start
+            coverage_end = pCoverageIndex[mate_bin_id].begin + vec_end
+            for i in xrange(coverage_index, coverage_end, 1):
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: I suspect that this can be optimized a little bit, but does not need to be part of this PR.

Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: It's also good to not use xrange, as python3 will not have it.

File: hicexplorer/hicBuildMatrix.py [hunk 2]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+            out_bam.write(mate2)
+
+    hic_matrix = coo_matrix((data, (row, col)), shape=(pMatrixSize, pMatrixSize), dtype='uint16')
+    row = []
+    col = []
+    data = []
+    if pOutputBamSet:
+        out_bam.close()
+
+    pQueueOut.put([hic_matrix, [one_mate_unmapped, one_mate_low_quality, one_mate_not_unique, dangling_end, self_circle, self_ligation, same_fragment,
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: Would it be more clear to create a class to store all these values and pass around the class?

File: hicexplorer/hicBuildMatrix.py [hunk 3]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+                                mate_not_close_to_rf, count_inward, count_outward,
+                                count_left, count_right, inter_chromosomal, short_range, long_range, pair_added, len(pMateBuffer1), pResultIndex, pCounter]])
+    return
+
+
+def write_bam(pOutputBam, pTemplate, pOutputFileBufferDir):
+
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: doc-string

File: hicexplorer/hicBuildMatrix.py [hunk 4]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+    out_bam = pysam.Samfile(pOutputFileBufferDir + pOutputBam, 'wb', template=pTemplate)
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: please use os.path.join() here.

File: hicexplorer/hicBuildMatrix.py [hunk 5]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+
+    counter = 0
+    while not os.path.isfile(pOutputFileBufferDir + 'done_processing') or os.path.isfile(pOutputFileBufferDir + str(counter) + '.bam_done'):
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: some here, os-path.join

File: hicexplorer/hicBuildMatrix.py [hunk 6]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+    Each mate pair is assessed to determine if it is
+    a valid Hi-C pair, in such case a matrix
+    reporting the counts of mates is constructed.
+
+    A bam file containing the valid Hi-C reads
+    is also constructed
+    """
+
+    # check pysam version
+    if LooseVersion(pysam.__version__) < LooseVersion("0.8.3"):
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: I would remove this from the code, the requirement.txt file should take care of this.

File: hicexplorer/hicBuildMatrix.py [hunk 7]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+
+    sys.stderr.write("reading {} and {} to build hic_matrix\n".format(args.samFiles[0].name,
+                                                                      args.samFiles[1].name))
+    str1 = pysam.Samfile(args.samFiles[0].name, 'rb')
+    str2 = pysam.Samfile(args.samFiles[1].name, 'rb')
+
+    args.samFiles[0].close()
+    args.samFiles[1].close()
+    outputFileBufferDir = args.outputFileBufferDir
+    if not outputFileBufferDir[:-1] == '/':
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: You can avoid this if you use os.path.join() everywhere.

File: hicexplorer/hicBuildMatrix.py [hunk 8]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+
+    matrix_size = len(bin_intervals)
+    bin_intval_tree = intervalListToIntervalTree(bin_intervals)
+    ref_id2name = str1.references
+
+    # build c_type shared memory for the interval tree
+    shared_array_list = []
+    index_dict = {}
+    j = 0
+    interval_list = []
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: not needed here, isn't it?

File: hicexplorer/hicBuildMatrix.py [hunk 9]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+    for i, seq in enumerate(bin_intval_tree):
+        start = j
+        interval_list = []
+        for interval in bin_intval_tree[seq]:
+            interval_list.append((interval.begin, interval.end, interval.data))
+            j += 1
+        end = j - 1
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: end is always len(bin_intval_tree[seq]) -1 + start isn't it?

File: hicexplorer/hicBuildMatrix.py [hunk 10]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+    # save the number of reads that overlap
+    # a bin.
+    # To save memory, coverage is not measured by bp
+    # but by bins of length 10bp
+    binsize = 10
+    number_of_elements_coverage = 0
+    start_pos_coverage = []
+    end_pos_coverage = []
+
+    for value in bin_intervals:
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: you can also write `for chrom, start, end in bin_intervals:`

File: hicexplorer/hicBuildMatrix.py [hunk 11]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+        chrom, start, end = value
+        start_pos_coverage.append(number_of_elements_coverage)
+
+        number_of_elements_coverage += (end - start) / binsize
+        end_pos_coverage.append(number_of_elements_coverage - 1)
+    pos_coverage_ = zip(start_pos_coverage, end_pos_coverage)
+    pos_coverage = RawArray(C_Coverage, pos_coverage_)
+    pos_coverage_ = None
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: Why is this needed?

File: hicexplorer/hicBuildMatrix.py [hunk 12]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+    all_threads_done = False
+    thread_done = [False] * args.threads
+    count_output = 0
+    count_call_of_read_input = 0
+    computed_pairs = 0
+    process_write_bam_file = Process(target=write_bam, kwargs=dict(pOutputBam=args.outBam.name, pTemplate=str1, pOutputFileBufferDir=outputFileBufferDir))
+    process_write_bam_file.start()
+    while not all_data_processed or not all_threads_done:
+
+        for i in xrange(args.threads):
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: range()

File: hicexplorer/hicBuildMatrix.py [hunk 13]
```diff
@@ -822,74 +844,363 @@ def main(args=None):
...
+                    pCoverageIndex=pos_coverage,
+                    pOutputFileBufferDir=outputFileBufferDir
+                ))
+                process[i].start()
+                count_output += 1
+                buffer_workers1[i] = None
+                buffer_workers2[i] = None
+
+            elif queue[i] is not None and not queue[i].empty():
+                result = queue[i].get()
```
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: This would be much cleaner if we can pass around some small classes or a Bunch()

File: hicexplorer/hicBuildMatrix.py (hunk 1 above)
Comment (joachimwolff) [inline hicexplorer/hicBuildMatrix.py]: I am not sure what to do with xrange vs. range. I know it is a potential issue to port the source code to Python 3 but on the other hand the behavior of 'range' in Python 2.7 is not wanted.

File: hicexplorer/hicBuildMatrix.py (hunk 2 above)
Comment (joachimwolff) [inline hicexplorer/hicBuildMatrix.py]: Maybe I use a sharedctype too.

File: hicexplorer/hicBuildMatrix.py (hunk 1 above)
Comment (bgruening) [inline hicexplorer/hicBuildMatrix.py]: use both :)
http://nullege.com/codes/search/six.moves.xrange

**Comment by bgruening** [comment]
@joachimwolff the test failing looks like a real error: https://travis-ci.org/maxplanck-ie/HiCExplorer/builds/225702496#L1796

File: hicexplorer/hicBuildMatrix.py [hunk 14]
```diff
@@ -490,139 +580,41 @@ def enlarge_bins(bin_intervals, chrom_sizes):
...
-    # check pysam version
-    from distutils.version import LooseVersion
-    if LooseVersion(pysam.__version__) < LooseVersion("0.8.3"):
-        exit("\n*ERROR*\n\nVersion of pysam has to be higher than 0.8.3. Current installed version is {}\n".format(pysam.__version__))
-
-    args = parse_arguments().parse_args(args)
-
-    # check that the log folder is valid
-    try:
-        QC.make_sure_path_exists(args.QCfolder)
```
Comment (joachimwolff) [inline hicexplorer/hicBuildMatrix.py]: You are right. I think it was getting lost when I did the manual merge with upstream.

File: hicexplorer/hicBuildMatrix.py (hunk 11 above)
Comment (joachimwolff) [inline hicexplorer/hicBuildMatrix.py]: To have the start and end positions for each chromosome in the 1D-Array?
Or do you mean that it could be written as: ```pos_coverage = RawArray(C_Coverage, zip(start_pos_coverage, end_pos_coverage))```

**Comment by joachimwolff** [comment]
With 7 cores and a buffer size of 400,000 RAM usage is now ~15 GB. There is some bug depending the 'lines per second' computation, the value which is displayed is to low. 

@bgruening  and @fidelram thanks a lot for the comments. I tried to implement them if possible.

File: hicexplorer/hicBuildMatrix.py (hunk 2 above)
Comment (joachimwolff) [inline hicexplorer/hicBuildMatrix.py]: @bgruening I tried that with hiding the variables in two classes, see commit 226c411. For some reason it is slower by a factor of 2. I don't want to invest more time to find the reason, I go back to the version that was used before. Maybe this can be done in a later PR.

**Comment by joachimwolff** [comment]
Based on performance troubles with the latest commits I decided to go back to a commit from a few days ago. I created for this a new branch to make sure the changes here are here if I ever need them. I close this PR and will open a new one.
